kernel/include/timeout_q.h

```diff
@@ -142,9 +142,9 @@ static inline int _abort_thread_timeout(struct k_thread *thread)
 	return _abort_timeout(&thread->base.timeout);
 }
 
+#ifdef CONFIG_KERNEL_DEBUG
```

Comment:
No, this one I don't like.  Peppering code with #if's is (much!) less readable than simply defining functions to be noops in some contexts.  What exactly does the MISRA rule say?

Note that this pattern is absolutely pervasive in Linux, and lots of other big C projects.

Reply:
Complains about dead code, so noops are considered dead code. In previous projects I also did noop implementations but none of them tried to be MISRA-C compliant.
In Zephyr we have both cases, so it's even worse, I don't think there is a workaround to avoid this, we may try to define a macro that expands to nothing when CONFIG_KERNEL_DEBUG is not defined, but then it'll be even more complicated.

Reply:
btw, if you are eager to have it on Zephyr, we can create a deviation for this since it is a required rule (not mandatory). But there are only 2 violations for this rule, my opinion is to have deviation for things that have more impact on code, either performance or readability. The memory allocation and VLAs are one example.

Reply:
Honestly we do this everywhere, take a look at wait_q.h for an example I'm literally reworking right now.  My guess is that the scanner is just too dumb to pick up the situations where the code doesn't "look" like a noop (e.g. it has an ARG_UNUSED() macro in it).  Or in other spots we use the preprocessor to #define the function away (which is IMHO worse as it lacks typesafety and we really want to be using empty functions).

If that's OK, then sure: I can arrange to have something in such functions such that they are not "empty" from the perspective of whatever analysis tools we are using. 

But I'm really not prepared to concede the core point: using empty functions that can be optimized out as a way to manage build time configuration **is a very big readability win**, and we don't want to drop that because of some dumb standard.

Reply:
the macros can trick the tool indeed. Ok, so moving forward on this:

1) Change the code to follow this pattern (using inline empty functions) and not have defined macros when the function is not available.

2) Document this deviation - at this point can be a description why we'll keep doing this, the arguments that you used are enough and I agree with you too.

3) list the cases where this happen

Can you help with this ?

Reply:
btw, I'll remove this commit from this pr.